### PR TITLE
add token tracking

### PR DESCRIPTION
**Pull request by NicoBliss**
adds positional data to the token struct that contains where that token starts
modified the lexer struct and made getc and ungetc wrappers to keep track of location
made the lexing function store tokens with positional data

### CODE REVIEW

**Comment by NicoBliss** [comment]
Would close issue #8

**Comment by NicoBliss** [comment]
last commit would close #32

File: src/lexer/token.h [hunk 1]
```diff
@@ -97,7 +97,9 @@ typedef enum {
...
 
 typedef struct {
-    TokenType type;              // What type of token this is.
-    char contents[TOKEN_LENGTH]; // The actual contents of the token.
-    unsigned length;             // How long the token is.
+    TokenType type;                 // What type of token this is.
+    char contents[TOKEN_LENGTH];    // The actual contents of the token.
+    unsigned length;                // How long the token is.
+    char source_file[TOKEN_LENGTH]; // The source file the token was in.
+    long position_in_file;           // Where in the file the token started, as an index.
```
Comment (adamhutchings) [inline src/lexer/token.h]: Is there a way to get row/column info from this? If not, we should maybe switch to that so we can give helpful error messages.

File: src/lexer/lex.c [hunk 2]
```diff
@@ -16,14 +18,26 @@ int in_string(char c, char s[]) {
...
 }
 
+int lexer_getchar(Lexer* l) {
+    l->position++;
+    l->buffer[0] = getc(l->fp);
+    return l->buffer[0];
+}
+
+int lexer_ungetchar(Lexer *l) {
+    // if called before getchar ever is this could cause problems
```
Comment (JakeRoggenbuck) [inline src/lexer/lex.c]: Maybe add an normal assert to prevent this and crash

Comment (JakeRoggenbuck) [reply src/lexer/lex.c]: assert(l->position >= 0);

File: src/lexer/token.h (hunk 1 above)
Comment (NicoBliss) [reply src/lexer/token.h]: my plan was to have the error reporting find that. It is simpler (i think?) to just crawl to the relevant position and hold onto a few rows of buffer. plus, errors are only usually called a few times, but this is tracked for every token

File: src/lexer/lex.c (hunk 2 above)
Comment (NicoBliss) [reply src/lexer/lex.c]: added!

File: src/lexer/token.h (hunk 1 above)
Comment (NicoBliss) [reply src/lexer/token.h]: ok so line and column information is now being tracked instead

**Review by adamhutchings — APPROVED** [review]
Beautiful!

**Pull request merged**